feat(content, reorder-group, header, footer, infinite-scroll, refresher): allow custom scroll target

sean-perkins opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help: https://ionicframework.com/docs/building/contributing -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures


## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying. -->

The following components are dependent on an implementation using `ion-content`, which is not compatible in implementations that use a custom scroll container, such as virtual-scroll.

Affected components:
- `ion-header` - fade effect and large titles (iOS)
- `ion-footer` - fade effect (iOS)
- `ion-infinite-scroll` - scrolling
- `ion-reorder-group` - ability to drag/reorder items
- `ion-refresher` - ability to pull to refresh

<!-- Issues are required for both bug fixes and features. -->
Issue URL:  #23437

## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

Developers can make use of a new global CSS class selector: `.ion-content-scroll-host` to specify to implementations using the above components/features, which container is the scroll target to attach listeners to. 

Developers using a custom scroll containers should additionally consider disabling scrolling on the host `ion-content` to prevent duplicate scroll containers (`scroll-y="false"` for vertical scrolling and `scroll-x="false"` for horizontal scrolling).

- `ion-header` fade effect can be used with a custom scroll container
- `ion-header` large/collapsible title can be used with a custom scroll container
- `ion-footer` fade effect can be used with a custom scroll container
- `ion-infinite-scroll` can be used with a custom scroll container (emitting `ionInfinite` when the container reaches threshold)
- `ion-reorder-group` can be used with a custom scroll container (allows reordering elements within target container)
- `ion-refresher` can be used with a custom scroll container (preserves iOS and MD animation effects without needing to specify a custom `#background-content` element)

All of the above features can be used against virtual scroll implementations.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

There were many areas of the implementation that were lacking e2e coverage of functionality, the following components had tests filled in, to cover the behavior changed with this PR: `ion-infinite-scroll`, `ion-refresher`, `ion-footer` (fade effect is exclusive to iOS).

sean-perkins commented:
@tlaverdure this will officially implement the functionality desired by your PR (#24426) with a static selector `.ion-content-scroll-host`. This selector name is still dependent on the team's review and could change if a better name is suggested. 

The intention of this feature is not to support ejecting from `ion-content` implementations, but it is possible with the resulting feature. If you decide to do so, you become responsible for all styling concerns that `ion-content` is currently handling for you as well as opinionated implementations such as the `#background-content` element for pull-to-refresh animations.

The intended use case is for better supporting virtual scroll with our components. With those implementations, you can disable the `ion-content` scroll container, but benefit from all internal styles and not needing to re-create the `#background-content` element for pull-to-refresh implementations. 

A dev build to test with is available: `6.0.10-dev.11646422799.5f745c5`.

I am going to close out your PR now that this feature is officially in our active work. I appreciate your time with that discovery as well as the dialog for refining the feature. If you have additional feedback or have an opportunity to test with the dev build and have concerns/issues, please let me know!

tlaverdure commented:
Thanks @sean-perkins! I'm going to test it out next week and let you know how it goes.

amandaesmith3 commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+  }));
+}
+
+/**
+ * Scrolls to a specific x/y location. If an `ion-content` is found, it will scroll
+ * using the public API `scrollToPoint` with a duration.
+ */
+export const scrollByPoint = (el: HTMLElement, x: number, y: number, durationMs: number) => {
+  if (el.tagName.toLowerCase() === ION_CONTENT_ELEMENT_SELECTOR) {
+    const content = el as HTMLIonContentElement;
+    return content.scrollByPoint(x, y, durationMs);
+  }
+  return Promise.resolve(el.scrollTo({
+    top: y,
+    left: x,
+    behavior: durationMs > 0 ? 'smooth' : 'auto'
+  }));
```
The comment references `scrollToPoint`, and the use of `scrollTo` would mirror that behavior, but the actual function is called and uses `scrollByPoint`, which is slightly different. (Check their implementations in `ion-content`.) It looks like `scrollByPoint` was your intention here since it's used in the `scroll-assist` shim, so the comment and fallback should be updated.

sean-perkins replied:
Very good catch! Updated the comment and the fallback implementation for the `HTMLElement`, to use scroll by point behavior.

sean-perkins commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+ */
+const ION_CONTENT_SELECTOR = `${ION_CONTENT_ELEMENT_SELECTOR}, ${ION_CONTENT_CLASS_SELECTOR}`;
+
+const hasOverflowScroll = (node: Element | null) => {
```
I would feel much more comfortable type casting from: 
```ts
const scrollContainer = getScrollContainer(el as HTMLElement);
```

and removing the type casting elsewhere. Ideally, functions should have narrow types of the explicit things they are dependent on. Otherwise they come highly coupled/harder to use over time. 

Datetime is an example of this, with the `DatetimeParts` type being required on all the utils, even if the util only needs a month to execute.

amandaesmith3 replied:
Works for me 👍

sean-perkins commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+  const overflowY = isElement && getComputedStyle(node).overflowY;
+  // Element is scrollable if there is overflow or if the overlay is explicit
+  const isScrollable = overflowY !== 'visible' && overflowY !== 'hidden';
+  // The element is the node if it's scrollable or the scroll height is larger than the client height
```
The element being checked by `hasOverflowScroll`, is a match/the element with overflow scroll if it is explicitly scrollable or if the scroll height is larger than the client height.

sean-perkins replied:
Ended up removing all this extra querying for scroll containers. Wasn't needed after our discussion around detaching ion-content 🎉

sean-perkins commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+export const getScrollElement = async (el: Element) => {
+  await new Promise(resolve => componentOnReady(el, resolve));
+
+  if (el?.tagName.toLowerCase() === ION_CONTENT_ELEMENT_SELECTOR) {
```
This check is very explicit to calling custom methods exposed on `ion-content`. I do not think we would want to open that up to other selectors, otherwise it will throw exceptions.

liamdebeasi commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+ * or a selector within the host, if supplied through `scrollTarget`.
+ */
+export const getScrollElement = async (el: Element) => {
+  await new Promise(resolve => componentOnReady(el, resolve));
```
We only need to do this if the element is `ion-content`. We should move this inside of the `if` block below.

liamdebeasi commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+  return false;
+}
+
+const getScrollContainer = (node: Element) => {
```
I don't get why we need to do all of this work. Since users are instructed to place `.ion-content-scroll-host` on their scrollable element, isn't querying for that class enough?

liamdebeasi replied:
I also have concerns over the usage of `getComputedStyle` in `hasOverflowScroll` as that will compute all of the styles for the element again.

sean-perkins replied:
Agreed, I think we can cut a lot of complexity here by removing this util and requiring the implementation to correctly place the class on the scrollable element. This was to help patch "detaching ion-content", but after our team discussion, we can remove those pieces.

liamdebeasi commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+ * using the public API `scrollToTop` with a duration.
+ */
+export const scrollToTop = (el: HTMLElement, durationMs: number): Promise<any> => {
+  if (el.tagName.toLowerCase() === ION_CONTENT_ELEMENT_SELECTOR) {
```
We do this check in multiple places. Can we make this a utility function instead?

liamdebeasi commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,130 @@
+    const content = el as HTMLIonContentElement;
+    return content.scrollByPoint(x, y, durationMs);
+  }
+  return Promise.resolve(el.scrollTo({
```
Can we use [Element.scrollBy](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollBy) instead?

sean-perkins commented:
@tlaverdure I wanted to update you on the status of this PR, based on decisions the team has discussed privately. 

The changes for querying that would support detaching from `ion-content`, ended up conflicting with existing Dx patterns we have implemented for `ion-refresher`, to require it to be used inside of `slot="fixed"`. Changing this behavior would be a breaking change and could also result in incorrect implementations of `ion-refresher` (specifically the iOS pull to refresh animation). 

As such, I have removed the referenced issue for the webkit performance bug. With reviewing that issue, there are a few items that stand out to me:
1. The performance overhead is an issue that Webkit will have to address with ShadowDOM. 
2. The reproduction use case does not use virtual scroll, a feature that could not fully be taken advantage of prior to this active PR. 

I would be happy to test virtual scroll against your reproduction to see if there are noticeable performance improvements that you can taken advantage of in the interim. 

I apologize for committing to handling your use case and the end result not addressing it. We will continue to evaluate improvements we can make that would incrementally improve the performance without causing a breaking change (or hold a larger improvement for v7). 

Thanks for your willingness to test this PR and your continued time in improving Ionic!

liamdebeasi commented on core/src/utils/content/content.utils.spec.ts:
```diff
@@ -0,0 +1,189 @@
+      tagName: 'MY-EL'
+    });
+
+    expect(consoleErrorMock).toHaveBeenCalledWith('<my-el> must be used inside ion-content or .ion-content-scroll-host.');
```
This is confusing. If I use `<my-el>` inside of `.ion-content-scroll-host`, I would expect the component to work as intended. The reality is that the component needs to be inside of `ion-content`. Whether or not you use `.ion-content-scroll-host` depends on if you are using virtual scroll.

liamdebeasi commented on core/src/utils/content/index.ts:
```diff
@@ -0,0 +1,98 @@
+ */
+const ION_CONTENT_SELECTOR = `${ION_CONTENT_ELEMENT_SELECTOR}, ${ION_CONTENT_CLASS_SELECTOR}`;
+
+const isIonContent = (el: Element) => el && el.tagName.toLowerCase() === ION_CONTENT_ELEMENT_SELECTOR;
```
Minor thing, but can we create an `ION_CONTENT_TAG_NAME` const so we do not have to do `toLowerCase` every time?

```js
const ION_CONTENT_TAG_NAME = 'ION-CONTENT';
el && el.tagName === ION_CONTENT_TAG_NAME;
```

liamdebeasi replied:
Minor thing, but can we create an `ION_CONTENT_TAG_NAME` const so we do not have to do `toLowerCase` every time?

```js
const ION_CONTENT_TAG_NAME = 'ION-CONTENT';

...

el && el.tagName === ION_CONTENT_TAG_NAME
```

liamdebeasi commented on core/src/components/header/test/scroll-target/e2e.ts:
```diff
@@ -0,0 +1,48 @@
+    describe('when the scroll container has overflow', () => {
+
+      it('should display the collapsed title on scroll', async () => {
+        const largeHeader = await page.find('ion-header[collapse="condense"]');
```
Can we take a screenshot once the large title has collapsed?

liamdebeasi replied:
Can we take screenshots?

liamdebeasi commented on core/src/components/header/test/scroll-target/e2e.ts:
```diff
@@ -0,0 +1,48 @@
+  describe('large title', () => {
+
+    it('should display the large title initially', async () => {
+      const largeHeader = await page.find('ion-header[collapse="condense"]');
```
Can we take screenshots?

liamdebeasi reviewed (changes requested):
Please be sure to also update the Virtual Scroll docs:

Vue: https://github.com/ionic-team/ionic-docs/blob/main/docs/vue/virtual-scroll.md#a-note-on-ionic-components
Angular: https://github.com/ionic-team/ionic-docs/blob/main/docs/angular/virtual-scroll.md#a-note-on-ionic-components
React: https://github.com/ionic-team/ionic-docs/blob/main/docs/react/virtual-scroll.md#a-note-on-ionic-components

(In particular, the "A Note on Ionic Components" section)

liamdebeasi reviewed (approved):
Good to merge once the other 4 points have been addressed. Great job on this 👍

liamdebeasi commented:
One more thing: Please make sure you update the commit subject to be more descriptive

sean-perkins commented:
PR to update the docs for virtual scrolling will be tracked here: https://github.com/ionic-team/ionic-docs/pull/2229

mariusbolik commented:
Hi @sean-perkins, thank you for the good work! Is there a dev build or any other way to test this feature?
